### PR TITLE
Delay Adapter#check_version to #configure_connection

### PR DESCRIPTION
**Pull request by casperisfine**
Ideally we should be able to checkout an Adapter instance without triggering a connection to the server.

For some adapters like Trilogy that is the case today, but only if you have a schema cache loaded, otherwise `check_version` will trigger a query.

I think this check can be delayed to when we actually use the connection.

### CODE REVIEW

**Review by rafaelfranca — APPROVED** [review]

**Comment by casperisfine** [comment]
There's some intricate build failures to investigate. I'll have another look tomorrow.

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 1]
```diff
@@ -672,7 +672,6 @@ def remove_connection_from_thread_cache(conn, owner_thread = conn.owner)
         def new_connection
           connection = Base.public_send(db_config.adapter_method, db_config.configuration_hash)
           connection.pool = self
```
Comment (composerinteralia) [inline activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:674]: We might be able to get rid of `connection.pool = self` here too, since we already do it later in `adopt_connection`. If I recall correctly it was added so we'd set up the schema cache before calling `check_version`.

**Review by composerinteralia — COMMENTED** [review]
👍🏻 Making this change was on my list of things to do. GitHub has a lingering Trilogy adapter patch that does basically the same thing, and it'll be nice for us to get that cleaned up.

**Comment by casperisfine** [comment]
I only have one failure left in Action Cable, but I'm having trouble reproducing it locally :/

**Comment by casperisfine** [comment]
I haven't had the time to figure out the Action Cable issue today, but I don't expect it to need much more change.

@matthewd any opinions here?

File: actioncable/lib/action_cable/subscription_adapter/postgresql.rb [hunk 2]
```diff
@@ -38,6 +38,7 @@ def with_subscriptions_connection(&block) # :nodoc:
           # will perform the necessary cleanup tasks
           ActiveRecord::Base.connection_pool.remove(conn)
         end
+        ar_conn.connect!
```
Comment (matthewd) [inline actioncable/lib/action_cable/subscription_adapter/postgresql.rb:41]: This feels surprising... `raw_connection` should probably take care of it?

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb [hunk 3]
```diff
@@ -378,6 +378,10 @@ def adapter_name
         self.class::ADAPTER_NAME
       end
 
+      def connected?
+        true
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb]: AbstractAdapter owns `@raw_connection`, so I think this can do the

```suggestion
        !@raw_connection.nil?
```

File: activerecord/lib/active_record/railtie.rb [hunk 4]
```diff
@@ -162,10 +162,13 @@ class Railtie < Rails::Railtie # :nodoc:
...
               end
-              next if current_version.nil?
 
-              if cache.schema_version != current_version
+              if current_version.nil?
+                connection_pool.schema_reflection.clear!
+                next
+              elsif cache.schema_version != current_version
                 warn "Ignoring #{filename} because it has expired. The current schema version is #{current_version}, but the one in the schema cache file is #{cache.schema_version}."
+                connection_pool.schema_reflection.clear!
```
Comment (matthewd) [inline activerecord/lib/active_record/railtie.rb:171]: This sounds reasonable but only semi-related... assuming it is separate (the instigating change?), might be better split out, it only because of how confusing it is that this is also about a different version check? 😅

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 5]
```diff
@@ -1036,6 +1040,8 @@ def configure_connection
           add_pg_decoders
 
           reload_type_map
+
+          super
```
[the remark this answers is not in the document]
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1044]: Unfortunately I had to move it down because we need `@type_map_for_results` to check the version...

File: activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb [hunk 6]
```diff
@@ -233,7 +233,7 @@ def with_connection
 
       # Returns true if a connection has already been opened.
       def connected?
-        synchronize { @connections.any? }
+        synchronize { @connections.any?(&:connected?) }
```
[the remark this answers is not in the document]
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:236]: Well, the doc says "connection has been opened". 

before my change this was true as soon as there one one adapter instance in the pool, now that the connection is properly lazy, it's no longer true. I had to change this to fix a good number of tests.

File: actioncable/lib/action_cable/subscription_adapter/postgresql.rb (hunk 2 above)
Comment (casperisfine) [reply actioncable/lib/action_cable/subscription_adapter/postgresql.rb:41]: Possibly, I was trying various stuff for fixing Action Cable, this probably isn't necessary, see: https://github.com/rails/rails/pull/49414

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (casperisfine) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb]: Ok, I refactored the PG adapter so that `@type_map_for_result` is initialized earlier. This now allow to do the version check on an unconfigured connection.

File: activerecord/lib/active_record/railtie.rb (hunk 4 above)
Comment (casperisfine) [reply activerecord/lib/active_record/railtie.rb:171]: Moved to https://github.com/rails/rails/pull/49415

**Comment by casperisfine** [comment]
I rebased on top of https://github.com/rails/rails/pull/49415, this solves the circular issue in PostgresAdapter.

File: activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb [hunk 7]
```diff
@@ -1182,7 +1191,6 @@ def add_pg_decoders
           coders.each { |coder| map.add_coder(coder) }
           @raw_connection.type_map_for_results = map
 
-          @type_map_for_results = PG::TypeMapByOid.new
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1185]: Is this change still required?

This method is called from `configure_connection`, which will be called multiple times over the instance lifetime (given reconnects). I think we want a new type map each time in that case -- but otherwise we probably need to at least clear it out.

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:1185]: Nope. It's a left over I need to remove
